java/core-arrow/src/main/java/sleeper/arrow/schema/SchemaBackedByArrow.java

```diff
@@ -0,0 +1,110 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sleeper.arrow.schema;
+
+
+import org.apache.arrow.vector.types.pojo.Field;
+
+import sleeper.core.schema.Schema;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static sleeper.arrow.schema.SchemaConverter.convertArrowSchemaToSleeperSchema;
+import static sleeper.arrow.schema.SchemaConverter.convertSleeperSchemaToArrowSchema;
+
+public class SchemaBackedByArrow {
+    private final org.apache.arrow.vector.types.pojo.Schema arrowSchema;
+    private final List<String> rowKeyFieldNames;
+    private final List<String> sortKeyFieldNames;
+    private final List<String> valueFieldNames;
+
+    private SchemaBackedByArrow(Builder builder) {
+        arrowSchema = Objects.requireNonNull(builder.arrowSchema, "arrowSchema must not be null");
+        rowKeyFieldNames = Objects.requireNonNull(builder.rowKeyFieldNames, "rowKeyFieldNames must not be null");
+        sortKeyFieldNames = Objects.requireNonNull(builder.sortKeyFieldNames, "sortKeyFieldNames must not be null");
+        valueFieldNames = arrowSchema.getFields().stream()
+                .map(Field::getName)
+                .filter(field -> !rowKeyFieldNames.contains(field) && !sortKeyFieldNames.contains(field))
+                .collect(Collectors.toList());
+        if (valueFieldNames.isEmpty()) {
+            throw new IllegalArgumentException("No value fields defined");
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static SchemaBackedByArrow fromSleeperSchema(Schema sleeperSchema) {
+        return builder()
+                .arrowSchema(convertSleeperSchemaToArrowSchema(sleeperSchema))
+                .sleeperSchema(sleeperSchema)
+                .build();
+    }
+
+    public static SchemaBackedByArrow fromArrowSchema(org.apache.arrow.vector.types.pojo.Schema arrowSchema,
+                                                      List<String> rowKeyFieldNames, List<String> sortKeyFieldNames) {
+        return builder()
+                .arrowSchema(arrowSchema)
+                .rowKeyFieldNames(rowKeyFieldNames)
+                .sortKeyFieldNames(sortKeyFieldNames)
+                .build();
+    }
+
+    public org.apache.arrow.vector.types.pojo.Schema getArrowSchema() {
+        return arrowSchema;
+    }
+
+    public Schema getSleeperSchema() {
+        return convertArrowSchemaToSleeperSchema(arrowSchema, rowKeyFieldNames, sortKeyFieldNames, valueFieldNames);
+    }
```

Comment:
It doesn't seem like we need to convert back to a Sleeper schema?

The was that we'd update places to use this class instead of a Sleeper schema. We want to be able to delete the Sleeper schema once everything is using this class.